[General] Add toggle to `[p]serverinfo` that shows more details

PredaaA opened the pull request:
### Type

- [ ] Bugfix
- [X] Enhancement
- [ ] New feature

### Description of the changes
Using `[p]serverinfo true` shows more details about the guild, such as verification levels, AFK channel and timeout, number of custom emojis, special and/or Nitro boost features. This is untoggled by default.

PredaaA committed:
Improvements of "serverinfo" command

- Added some more informations about the server.
  (Verification level, AFK channel and timeout, custom emojis and special features).
- Added custom texts for Region (and Verification levels).
- Added more details about members (Number of humans, bots, and all status).
- Show special features only if the server has one.
- Show "Verified" logo only if the server is verified.
- And changed the footer by adding how many time the bot as join the server.

mikeshardmind commented on redbot/cogs/general/general.py:
```diff
@@ -218,35 +218,93 @@ def __init__(self):
         except discord.Forbidden:
             await ctx.send(_("I need the `Embed links` permission to send this."))
 
+    async def check(self, true):
```
This doesn't need to be a coroutine. It also doesn't need to belong to the cog class.

mikeshardmind commented on redbot/cogs/general/general.py:
```diff
@@ -218,35 +218,93 @@ def __init__(self):
+                value=_("{vip} VIP Regions\n{van} Vanity URL\n{splash} Splash Invite\n{m_emojis} More Emojis\n{verify} Verified").format(
+                    vip=await self.check("VIP_REGIONS" in guild.features), van=await self.check("VANITY_URL" in guild.features),
+                    splash=await self.check("INVITE_SPLASH" in guild.features), m_emojis=await self.check("MORE_EMOJI" in guild.features),
+                    verify=await self.check("VERIFIED" in guild.features)))
```
This doesn't need to be calling a coroutine to do this. you can define functions inside of functions.

So for this, above `data.add_field` you could have
```py
def check_feature(feature):
    return "\N{WHITE HEAVY CHECK MARK}" if feature in guild.features else "\N{CROSS MARK}"

format_kwargs = {
    "vip": check_feature("VIP_REGIONS"),
    "van": check_feature("VANITY_URL"),
    ...
}

value=_(
    "{vip} VIP Regions\n{van} Vanity URL\n{splash} Splash Invite\n{m_emojis} More Emojis\n{verify} Verified"
).format(**format_kwargs)
```

This has some added benefits for readability of intent as well.

mikeshardmind commented on redbot/cogs/general/general.py:
```diff
@@ -218,35 +218,93 @@ def __init__(self):
+            data.add_field(name=_("Members :"),
+            value=_(f"Total users : **{total_users}**\nHumans : **{humans}** • Bots : **{bots}**\n📗 `{online}` 📙 `{idle}` 📕 `{dnd}` 📓 `{offline}`"))
+        data.add_field(name=_("Channels :"), value=_(f"💬 Text : **{text_channels}**\n🔊 Voice : **{voice_channels}**"))
+        data.Empty
```
`data.Empty` 
This line can be removed.

mikeshardmind commented on redbot/cogs/general/general.py:
```diff
@@ -218,35 +218,93 @@ def __init__(self):
+        data.Empty
+        data.add_field(name=_("Utility :"), value=_(f"Owner : **{guild.owner}**\nRegion : **{region[str(guild.region)]}**\nVerif. level : **{verif[int(guild.verification_level)]}**\nServer ID : **{guild.id}**"))
+        data.add_field(name=_("Misc :"), value=_(f"AFK channel : **{guild.afk_channel}**\nAFK Timeout : **{guild.afk_timeout}sec**\nCustom emojis : **{len(guild.emojis)}**\nRoles : **{len(guild.roles)}**"))
+        if guild.features != []:
```
```suggestion
        if guild.features:
```

empty lists evaluate as false.

Zephyrkul commented:
Just FYI, your PR is failing a [style check (clicky).](https://github.com/Cog-Creators/Red-DiscordBot/blob/V3/develop/.github/CONTRIBUTING.md#43-style)

mikeshardmind commented on redbot/cogs/general/general.py:
```diff
@@ -218,35 +218,93 @@ def __init__(self):
-
+        if total_users > 8000:
+            data.add_field(name=_("Members :"),
+            value=_(f"Total users : **{total_users}**\nHumans : **{humans}** • Bots : **{bots}**\n📗 `{online}` 📙 `{idle}`\n📕 `{dnd}` 📓 `{offline}`"))
```
cant mix f-string with i18n translation like this, use `.format` (again, and at each additional instance)

mikeshardmind commented on redbot/cogs/general/general.py:
```diff
@@ -218,35 +218,93 @@ def __init__(self):
+        joined_at = guild.me.joined_at
+        since_joined = (ctx.message.created_at - joined_at).days
+        bot_joined = joined_at.strftime("%d %b %Y %H:%M:%S")
+        joined_on = (_(f"{ctx.bot.user.name} joined this server on {bot_joined}. That's over {since_joined} days ago !"))
```
cant mix f-string with i18n translation like this, use  `.format`

mikeshardmind reviewed (changes requested):
Overall nice looking embed with more info. Would love to see some leveraging of the mobile status info in this at some point too. There are a few things which should be changed here before this gets merged.

PredaaA committed:
Black formatting for serverinfo

PredaaA committed:
Black formatting fix and Sinbad suggestions

PredaaA committed:
Fix the discord.errors.NotFound on server command

Added exception on server command for the phrase "To leave a server, just type its number.".

PredaaA committed:
Update core_commands.py

PredaaA committed:
Delete core_commands.py

PredaaA committed:
Create core_commands.py

PredaaA committed:
Update core_commands.py

PredaaA committed:
Update core_commands.py

PredaaA committed:
Merge pull request #1 from PredaaA/PredaaA-patch-1

Update core_commands.py

PredaaA committed:
Delete core_commands.py

PredaaA committed:
Create core_commands.py

PredaaA committed:
Update core_commands.py

PredaaA committed:
Update core_commands.py

PredaaA committed:
Little changes on serverinfo command

- Adding "lurkers" check, for the new feature that are on bot which allow users to join a server in read-only (Works only on verified servers for now)
- Adding streaming status.
- Adding mobile status with suggestion of Sinbad.

PredaaA committed:
Black style fix

PredaaA commented:
I know that not everyone likes my PR for this serverinfo command, and I really can understand that because it's totally different than the original command, but also some people like it so that's why I continue to trying to improve it.

- Fixed all changes that Sinbad asked.
- Adding "lurkers" check, for the new feature that are on Discord which allow users to join a server in read-only (Works only on verified servers for now, only beta)
- Adding streaming status.
- Adding mobile status that a suggestion from Sinbad.

Thanks to testing channel on Red Discord for all your help.
(I'm sorry for all commits you see above, it wasn't intended to show it here, really sorry for that)

Zephyrkul commented:
If this PR ends up being rejected, I wouldn't mind seeing it in a 3rd party cog.

PredaaA commented:
That's what I've already done. I've added it into a 3rd party cog in my repo, that replace the original command.

Tobotimus commented:
The embed is more cluttered, but I think the information on it is still easy to find. I was initially thinking of making it an option / add a "verbosity" argument to serverinfo, but I don't think that's necessary. So purely from a black box perspective I'd give this my :+1:

Kowlin commented:
I feel like this would be an appropriate candidate for a ``-verbose`` flag.

mikeshardmind commented:
The detection of lurkers there is a bug to be fixed in discord (see: https://github.com/discordapp/discord-api-docs/issues/855) , we shouldn't include something which exposes them in the mean time.

I also agree with kowlin about the verbosity setting, and it does not have to significantly change the logic to use this, just place certain parts under and `if verbose:` statement.

PredaaA committed:
Add verbose for serverinfo and disable lurkers

I've added a bool that are False by default and sent the original serverinfo command, and if set to True send my edited version.
And I've disabled the lurkers detection until this bug (https://github.com/discordapp/discord-api-docs/issues/855) isn't fixed.

PredaaA commented:
I've added a bool that are False by default and sent the original serverinfo command, and if set to True send my edited version, and finally I think it's better with that, if users want more information about the server it just need to set True when send the command.
I've just keep the personalized text for region in original one, in my opinion it looks better like this, but maybe I'm wrong, tell me if you think it's better without.

And I've disabled the lurkers detection until the bug @mikeshardmind mentionned isn't fixed.

PredaaA committed:
Add of India region

PredaaA committed:
Code cleanup

PredaaA committed:
Emojis are not needed to be translated

Tobotimus committed:
Merge branch 'V3/develop' into V3/develop

mikeshardmind commented on redbot/cogs/general/general.py:
```diff
@@ -220,31 +220,164 @@ def __init__(self):
+        offline = len([m.status for m in guild.members if m.status == discord.Status.offline])
+        streaming = len([m for m in guild.members if isinstance(m.activity, discord.Streaming)])
+        mobile = len([m for m in guild.members if m.is_on_mobile()])
+        lurkers = len([m for m in guild.members if m.joined_at is None])
```
Once this line is removed (and further uses of the var are gone) this is still a change we can add if you're still interested in doing this. (the lurkers being exposed like this was an API bug when it was happening (they aren't anymore) and there are other cases where joined_at can be None)

PredaaA replied:
I just updated the PR with all changes that I've done since last time I touched that PR.

PredaaA committed:
Update serverinfo

PredaaA committed:
Changelog.

PredaaA committed:
Merge branch 'V3/develop' into V3/develop

PredaaA committed:
Update 2382.enhancement.rst

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+        """
+        Show server information.
+    
+        `details`: Toggle it to `True` to show more informations about this server.
```
```suggestion
        `details`: Toggle it to `True` to show more information about this server.
```

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+        """
         guild = ctx.guild
+        passed = (ctx.message.created_at - guild.created_at).days
+        created_at = _("Since {date}. That's over {num} days ago!").format(
```
I think one of these would be clearer:
```suggestion
        created_at = _("Active since {date}. That's over {num} days ago!").format(
```
```suggestion
        created_at = _("Running since {date}. That's over {num} days ago!").format(
```
```suggestion
        created_at = _("Created on {date}. That's over {num} days ago!").format(
```

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+                "🎥": lambda x: x.activity == discord.Streaming,
+                "📱": lambda x: x.is_on_mobile(),
+            }
+            member_msg = _("Total Users: {}\n").format(bold(f"{online}/{total_users}"))
```
I think it's gonna be clearer to call this "Users online". I also put `{online}/{total_users}` cause it might help translators to know the context.
```suggestion
            member_msg = _("Users online: **{online}/{total_users}**\n").format(
                online=online, total_users=total_users
            )
```

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+                "MEMBER_LIST_DISABLED": _("Member list disabled"),
+            }
+            guild_features_list = [
+                f"✅ {name}" for feature, name in features.items() if feature in set(guild.features)
```
Nitpick: there's no need to create a new set on each iteration.
```suggestion
                f"✅ {name}" for feature, name in features.items() if feature in guild.features
```

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+                url=guild.icon_url
+                if guild.icon_url
+                else "https://cdn.discordapp.com/embed/avatars/1.png"
+            )
```
I think it's better to just not show server icon at all if it's not set instead of using unrelated one.
```suggestion
            if guild.icon_url:
                data.set_thumbnail(url=guild.icon_url)
```

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+                    "Owner: {owner}\nRegion: {region}\nVerif. level: {verif}\nServer ID: {id}"
+                ).format(
+                    owner=bold(str(guild.owner)),
+                    region=bold(vc_regions[str(guild.region)]),
```
In case there's gonna be added a new voice region, this will make the command less prone to errors and just show "unfriendly" name of the voice region if we don't have a friendly one in the dictionary.

I also replaced `bold()` call with wrapping in `**` as `bold()` also escapes the given text which doesn't work great with underline that is added to VIP regions.
```suggestion
                    region=f"**{vc_regions.get(str(guild.region)) or str(guild.region)}**",
```

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+                    else bold(_("Not set")),
+                    afk_timeout=bold(humanize_timedelta(seconds=guild.afk_timeout)),
+                    emojis=bold(humanize_number(len(guild.emojis))),
+                    roles=bold(humanize_number(len(guild.roles))),
```
Nitpick, small i18n improvement.
```suggestion
                    "AFK channel: {afk_chan}\nAFK Timeout: {afk_timeout}\nCustom emojis: {emoji_count}\nRoles: {role_count}"
                ).format(
                    afk_chan=bold(str(guild.afk_channel))
                    if guild.afk_channel
                    else bold(_("Not set")),
                    afk_timeout=bold(humanize_timedelta(seconds=guild.afk_timeout)),
                    emoji_count=bold(humanize_number(len(guild.emojis))),
                    role_count=bold(humanize_number(len(guild.roles))),
```

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+                "📕": lambda x: x.status == discord.Status.idle,
+                "📓": lambda x: x.status == discord.Status.offline,
+                "🎥": lambda x: x.activity == discord.Streaming,
+                "📱": lambda x: x.is_on_mobile(),
```
I think that circles will more clearly indicate the meaning of the statuses they represent:
```suggestion
                "\N{LARGE GREEN CIRCLE}": lambda x: x.status == discord.Status.online,
                "\N{LARGE ORANGE CIRCLE}": lambda x: x.status == discord.Status.idle,
                "\N{LARGE RED CIRCLE}": lambda x: x.status == discord.Status.do_not_disturb,
                "\N{MEDIUM WHITE CIRCLE}": lambda x: x.status == discord.Status.offline,
                "\N{LARGE PURPLE CIRCLE}": lambda x: x.activity == discord.Streaming,
                "\N{MOBILE PHONE}": lambda x: x.is_on_mobile(),
```
You also used idle status twice, fixed that in suggestion as well.

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+                "none": _("0 - None"),
+                "low": _("1 - Low"),
+                "medium": _("2 - Medium"),
+                "high": _("3 - Hard"),
```
```suggestion
                "high": _("3 - High"),
```

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+                "VANITY_URL": _("Vanity URL"),
+                "MORE_EMOJI": _("More Emojis"),
+                "COMMERCE": _("Commerce"),
+                "LURKABLE": _("Lurkable"),
```
LURKABLE no longer exists
```suggestion
                "COMMERCE": _("Commerce"),
```

Jackenmen commented on redbot/cogs/general/general.py:
```diff
@@ -234,38 +240,210 @@ def __init__(self):
+                "PARTNERED": _("Partnered"),
+                "VERIFIED": _("Verified"),
+                "DISCOVERABLE": _("Server Discovery"),
+                "PUBLIC": _("Public"),
```
There's one flag that is not included in this list, "PUBLIC_DISABLED" (see https://discordapp.com/developers/docs/resources/guild#guild-object-guild-features)
```suggestion
                "PUBLIC": _("Public"),
                "PUBLIC_DISABLED": _("Public disabled"),
```

PredaaA committed:
Adress Jack's requests changes.

PredaaA committed:
oops

Jackenmen committed:
Put guild description first + few last nitpicks

Jackenmen reviewed (approved):
Looks good to me, thanks for the PR!
